File: elixir/lib/2015/01_not_quite_lisp/aoc201501.ex
defmodule AOC2015.Day01 do
  @moduledoc """
  Advent of Code 2015, day 1: Not Quite Lisp
  """
  require AOC

  @steps %{"(" => 1, ")" => -1}

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("", trim: true)
    |> Enum.map(fn c -> @steps[c] end)
  end

  @doc """
  Solve part 1
  """
  def part1(input) do
    input |> Enum.sum()
  end

  @doc """
  Solve part 2
  """
  def part2(input) do
    1 + (input |> Enum.scan(&(&1 + &2)) |> Enum.find_index(&(&1 < 0)))
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2015/01_not_quite_lisp/aoc201501.py
"""AoC 1, 2015: Not Quite Lisp"""

# Standard library imports
import itertools
import pathlib
import sys

STEPS = {"(": 1, ")": -1}


def parse(puzzle_input):
    """Parse input"""
    return [STEPS[paren] for paren in puzzle_input]


def part1(data):
    """Solve part 1"""
    return sum(data)


def part2(data):
    """Solve part 2"""
    steps = enumerate(itertools.accumulate(data), start=1)
    return next(itertools.dropwhile(lambda d: d[1] >= 0, steps))[0]


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
